src/core/common/trickle_timer.cpp

```diff
@@ -53,6 +53,62 @@ TrickleTimer::TrickleTimer(Instance &aInstance, Handler aHandler)
 {
 }
 
+TimeMilli TrickleTimer::GetLastTimerStart(void)
+{
+    TimeMilli aStartTime = TimerMilli::GetFireTime();
+
+    switch (mPhase)
+    {
+    case kBeforeRandomTime:
+        aStartTime -= mTimeInInterval;
+        break;
+
+    case kAfterRandomTime:
+        aStartTime -= mInterval - mTimeInInterval;
+        break;
+    }
+
+    return aStartTime;
+}
+
+void TrickleTimer::SetIntervalMax(uint32_t aIntervalMax)
+{
+    OT_ASSERT(aIntervalMax >= mIntervalMin);
+
+    VerifyOrExit(aIntervalMax != mIntervalMax);
+
+    mIntervalMax = aIntervalMax;
+
+    if (aIntervalMax < mInterval)
+    {
+        // the new interval is smaller than the running interval, so re-configure the timer to
+        // fire at the maximum value
+
+        TimeMilli newFireTime = GetLastTimerStart();
+        newFireTime += aIntervalMax;
+
+        // In plain mode, just fire at the sooner time and
+        // the handler will take care of everything else.
+        //
+        // In trickle mode, fire at a sooner time, but we also
+        // need to manipulate mInterval and mTimeInInterval.
+        //
+        // From what I can tell, even if newFireTime is less
+        // than now, it will fire.
+
+        mInterval = aIntervalMax;
+        if (mTimeInInterval > aIntervalMax)
+        {
+            mTimeInInterval = aIntervalMax;
+        }
+
+        TimerMilli::FireAt(newFireTime);
```

Comment:
If I am not mistaken, (under `mMode == kModeTrickle`)  if we have `mPhase == kBeforeRandomTime` and `mTimeInInterval < aIntervalMax` then the new calculated `newFireTime` may be incorrect. 

An example:

```
    |<------mInterval = 20 sec --------------------------->|
    |                                         |            | 
    |<------- new aIntervalMax = 15 sec ----->|            |
    |                              |          |            |
    |<-- mTimeInInterval=11 sec -->|          |            |
    |                              |                       | 
                 ^
                 Now 
```

- `mInterval = 20 sec`, `mTimeInInterval = 11 sec`,  new `aIntervalMax = 15 sec`
- Since `mTimeInInterval < aIntervalMax` we keep `mTimeInInterval` as is (will remain `11`).
- We set the `mInterval` to `15`.
- Since we are in `mPhase == kBeforeRandomTime`,  `GetLastTimerStart()` will return the start of current trickle interval
- `newFireTime` will point to 15 sec after the start of trickle interval.
- When the timer fires, we are still in `kBeforeRandomTime` phase so we do the following:
```c++
        case kBeforeRandomTime:
            // We reached end of random `mTimeInInterval` (aka `t`)
            // within the current interval. Trickle timer invokes
            // handler if and only if the counter is less than the
            // redundancy constant.

            mPhase = kAfterRandomTime;
            TimerMilli::Start(mInterval - mTimeInInterval);
            VerifyOrExit(mCounter < mRedundancyConstant);
            break;
```
Which then start the timer for remaining time to end of current interval calculated as `mInterval - mTimeInInterval = 15 - 11 = 4`, so the overall interval becomes `15 + 4 = 19` second long which is not what we want.

I think in this case, 
- The timer is already scheduled to fire at `mTimeInInterval` after start (or time `t`) which we keep the same, so we don't need to change the timer firetime.
- We only need to update the `mInterval` to `aIntervalMax` so that after timer fires, we apply the shorter interval.

Reply:
Thanks. Looks good.  👍 

I think `ii.b` is possible, like this:

```
    |<------ mInterval = 30--------------------------->|
    |<-- mTimeInInterval=18 sec -->|       ^           |                    
    |                              |       ^           |                 
    |<-- aMaxInterval = 10 ->|     |       ^           |
    |                                      ^           |
                                           Now
```
But I think we need to do the same in this case as in `ii.b` change `mInterval` and re-schedule to fire `aIntervalMax` from last interval start (which will then fire immediately)

Reply:
We may be able to combine the case.

- The only case that seem to be different is `i.b`.

- In all other cases we can do the same set of things: 
  - Change `mInterval = aIntervalMax` 
  - Change `mTimeInInterval = aIntervalMax` (it only matters in `i.a` case,  ignored in other cases and harmless to change)
  - And re-schedule timer to fire `aIntervalMax` from the last start interval.
  - And we keep `mPhase` as is.

Does this make sense?

Reply:
Thanks for the diagram for `ii.a`, that makes it clear that it is possible. As you say, taking the same approach as `ii.b` should work.

I think that even in `i.b` we need to set `mInterval` = `aIntervalMax` (according to the logic above)

So with all of this worked through, I think the only change necessary is to update `GetLastTimerStart()` to `GetLastIntervalStart()`